lib/grpc_generator.dart

```diff
@@ -0,0 +1,252 @@
+  void _addDependency(GenerationContext ctx, String fqname, String location) {
+    if (_deps.containsKey(fqname)) return; // Already added.
+
+    MessageGenerator mg = ctx.getFieldType(fqname);
+    if (mg == null) {
+      _undefinedDeps[fqname] = location;
+      return;
+    }
+    mg.checkResolved();
+    _deps[mg.fqname] = mg;
+  }
+
+  /// Adds dependencies of [generate] to [imports].
+  ///
+  /// For each .pb.dart file that the generated code needs to import,
+  /// add its generator.
+  void addImportsTo(Set<FileGenerator> imports) {
+    for (var mg in _deps.values) {
+      imports.add(mg.fileGen);
+    }
+  }
+
+  /// Returns the Dart class name to use for a message type.
+  ///
+  /// Throws an exception if it can't be resolved.
+  String _getDartClassName(String fqname) {
+    var mg = _deps[fqname];
+    if (mg == null) {
+      var location = _undefinedDeps[fqname];
+      throw 'FAILURE: Unknown type reference (${fqname}) for ${location}';
+    }
+    if (fileGen.package == mg.fileGen.package || mg.fileGen.package == "") {
+      // It's either the same file, or another file with the same package.
+      // (In the second case, we import it without using "as".)
+      return mg.classname;
+    }
+    return mg.packageImportPrefix + "." + mg.classname;
+  }
+
+  void generate(IndentingWriter out) {
+    _generateClient(out);
+    out.println();
+    _generateService(out);
+  }
+
+  void _generateClient(IndentingWriter out) {
+    out.addBlock('class $_clientClassname {', '}', () {
+      out.println('final ClientChannel _channel;');
+      out.println();
+      for (final method in _methods) {
+        method.generateClientMethodDescriptor(out);
+      }
+      out.println();
+      out.println('$_clientClassname(this._channel);');
+      for (final method in _methods) {
+        method.generateClientStub(out);
+      }
+    });
+  }
+
+  void _generateService(IndentingWriter out) {
+    out.addBlock('abstract class $_serviceClassname extends Service {', '}',
+        () {
+      out.println('String get \$name => \'$_fullServiceName\';');
+      out.println();
+      out.addBlock('$_serviceClassname() {', '}', () {
+        for (final method in _methods) {
+          method.generateServiceMethodRegistration(out);
+        }
+      });
+      out.println();
+      for (final method in _methods) {
+        method.generateServiceMethodPreamble(out);
+      }
+      for (final method in _methods) {
+        method.generateServiceMethodStub(out);
+      }
+    });
+  }
+}
+
+class _GrpcMethod {
+  final MethodDescriptorProto _method;
+
+  final GrpcServiceGenerator _service;
+
+  String _grpcName;
+  String _dartName;
+  String _serviceName;
+
+  bool _clientStreaming;
+  bool _serverStreaming;
+
+  String _requestType;
+  String _responseType;
+
+  String _argumentType;
+  String _clientReturnType;
+  String _serverReturnType;
+
+  _GrpcMethod(this._service, this._method) {
```

Comment:
Given that this class isn't subclassed, you could consider having a simpler constructor that just assigns to fields and a factory method that does the work (substring, etc.). That would allow you to make more fields final too.